File: internal/mq/topics.go
package mq

import "time"

// ── Topic constants ───────────────────────────────────────────────────────────
// Single source of truth for all MQ topic strings used across the codebase.
// Mirrored in internal/ui/assets/js/mq/topics.js — keep both in sync.
const (
	// Peer lifecycle — published locally by run.go PeerTable bridge.
	TopicPeerAnnounce = "peer:announce"
	TopicPeerGone     = "peer:gone"

	// Call signaling — P2P between peers; hangup also published locally by routes/call.go.
	TopicCallPrefix = "call:" // + channelID

	// Call loopback ICE — Go → browser only (Phase 4).
	// Published by PublishLoopbackICE when Go's LocalPC generates ICE candidates.
	// Browser subscribes via Goop.mq.onLoopbackICE(channelId, fn).
	// Topic format: "call:loopback:" + channelID
	TopicCallLoopbackPrefix = "call:loopback:" // + channelID

	// Group protocol — P2P; group.invite is not scoped to a group ID.
	TopicGroupPrefix = "group:"  // + groupID + ":" + type
	TopicGroupInvite = "group.invite"

	// Listen state — published locally by listen.Manager.
	TopicListenPrefix = "listen:" // + groupID + ":state"

	// Chat — direct P2P and broadcast to all peers.
	TopicChat          = "chat"
	TopicChatBroadcast = "chat.broadcast"

	// Internal MQ event log — published locally by mq.logMQEvent.
	TopicLogMQ = "log:mq"
)

// ── Call signal type constants ─────────────────────────────────────────────────
// Value of the "type" field inside all call:* message payloads.
// Used in both browser mode (video-call.js) and native mode (call-native.js / Go/Pion).
const (
	CallTypeRequest    = "call-request"  // caller → callee: initiate a call
	CallTypeAck        = "call-ack"      // callee → caller: call accepted, SDP exchange starts
	CallTypeOffer      = "call-offer"    // caller → callee: SDP offer (after ack)
	CallTypeAnswer     = "call-answer"   // callee → caller: SDP answer
	CallTypeICE        = "ice-candidate" // either → other: trickle ICE candidate
	CallTypeHangup     = "call-hangup"   // either side: end the call
	CallTypeLoopbackICE = "loopback-ice" // Go → browser: LocalPC ICE candidate (Phase 4)
)

// ── Payload structs ───────────────────────────────────────────────────────────

// ── Call signal payloads ── topic: "call:{channelID}" ─────────────────────────
//
// All call signals share the topic "call:{channelID}" and are routed by the
// "type" field. Both browser (video-call.js) and native (call-native.js/Go Pion)
// modes use these same payload shapes.
//
// P2P signaling sequence:
//
//   caller                          callee
//   ──────────────────────────────────────────────────────────────
//   sendCallRequest ────────────────► (incoming call modal)
//                   ◄──────────────── sendCallAck  (on accept)
//   sendCallOffer   ────────────────►
//                   ◄──────────────── sendCallAnswer
//   sendCallICE ◄──────────────────► sendCallICE  (trickle, both ways)
//   sendCallHangup  ────────────────► (or either side, any time)
//
// Native-mode note: after the initial call-request/ack, Go/Pion handles the
// SDP offer/answer and peer-to-peer ICE exchange internally. The browser
// receives the media stream via WebSocket (/api/call/media/{channel}) using
// WebM/MSE (Phase 4). Browser ↔ local-Go loopback ICE uses the separate
// "call:loopback:{channelID}" topic (Phase 4).

// CallRequestPayload is sent by the caller to invite the remote peer.
type CallRequestPayload struct {
	Type        string `json:"type"`                  // CallTypeRequest
	Constraints any    `json:"constraints,omitempty"` // browser-mode media constraints
}

// CallAckPayload is sent by the callee after accepting the call.
// Receipt triggers the caller to create and send the SDP offer.
type CallAckPayload struct {
	Type string `json:"type"` // CallTypeAck
}

// CallOfferPayload carries the SDP offer from the caller to the callee.
type CallOfferPayload struct {
	Type string `json:"type"` // CallTypeOffer
	SDP  string `json:"sdp"`
}

// CallAnswerPayload carries the SDP answer from the callee back to the caller.
type CallAnswerPayload struct {
	Type string `json:"type"` // CallTypeAnswer
	SDP  string `json:"sdp"`
}

// CallICECandidateInit is the standard RTCIceCandidateInit shape (W3C WebRTC).
type CallICECandidateInit struct {
	Candidate     string `json:"candidate"`
	SDPMid        string `json:"sdpMid,omitempty"`
	SDPMLineIndex uint16 `json:"sdpMLineIndex"`
}

// CallICEPayload carries a trickle ICE candidate between peers.
type CallICEPayload struct {
	Type      string               `json:"type"`      // CallTypeICE
	Candidate CallICECandidateInit `json:"candidate"`
}

// CallLoopbackICEPayload is published locally (Go → browser) by PublishLoopbackICE.
// The browser adds this candidate to its loopback RTCPeerConnection (Phase 4).
// Topic: "call:loopback:{channelID}" (TopicCallLoopbackPrefix + channelID)
type CallLoopbackICEPayload struct {
	Type      string               `json:"type"`       // CallTypeLoopbackICE
	ChannelID string               `json:"channel_id"`
	Candidate CallICECandidateInit `json:"candidate"`
}

// ── Peer lifecycle payloads ─────────────────────────────────────────────────────

// PeerAnnouncePayload is the payload for TopicPeerAnnounce.
// Published by run.go whenever the PeerTable emits an "update" event.
type PeerAnnouncePayload struct {
	PeerID         string    `json:"peerID"`
	Content        string    `json:"content"`
	Email          string    `json:"email,omitempty"`
	AvatarHash     string    `json:"avatarHash,omitempty"`
	VideoDisabled  bool      `json:"videoDisabled,omitempty"`
	ActiveTemplate string    `json:"activeTemplate,omitempty"`
	Verified       bool      `json:"verified,omitempty"`
	Reachable      bool      `json:"reachable"`
	Offline        bool      `json:"offline"`
	LastSeen       int64     `json:"lastSeen"` // Unix milliseconds
	Favorite       bool      `json:"favorite,omitempty"`
	// Internal only — not sent to browser, populated from CachedPeer after lookup.
	LastSeenTime time.Time `json:"-"`
}

// PeerGonePayload is the payload for TopicPeerGone.
// Published by run.go when a peer is pruned from the in-memory PeerTable.
type PeerGonePayload struct {
	PeerID string `json:"peerID"`
}

// CallHangupPayload is the payload published locally by routes/call.go
// when a native call session's HangupCh fires.
type CallHangupPayload struct {
	Type      string `json:"type"`        // always "call-hangup"
	ChannelID string `json:"channel_id"`
}

// ── Typed publish helpers ─────────────────────────────────────────────────────

// PublishPeerAnnounce pushes a peer metadata update to the browser via MQ SSE.
func (m *Manager) PublishPeerAnnounce(p PeerAnnouncePayload) {
	m.PublishLocal(TopicPeerAnnounce, "", p)
}

// PublishPeerGone notifies the browser that a peer has been pruned from the table.
func (m *Manager) PublishPeerGone(peerID string) {
	m.PublishLocal(TopicPeerGone, "", PeerGonePayload{PeerID: peerID})
}

// PublishCallHangup notifies the browser that a native call session has ended.
// Called by routes/call.go watchHangup() when sess.HangupCh() fires.
func (m *Manager) PublishCallHangup(channelID string) {
	m.PublishLocal(TopicCallPrefix+channelID, "", CallHangupPayload{
		Type:      CallTypeHangup,
		ChannelID: channelID,
	})
}

// PublishLoopbackICE pushes a Go LocalPC ICE candidate to the browser.
// Called by routes/call.go during Phase 4 loopback negotiation when Go's
// local PeerConnection generates an ICE candidate for the browser's loopback PC.
// Topic: "call:loopback:{channelID}" — browser subscribes via Goop.mq.onLoopbackICE.
func (m *Manager) PublishLoopbackICE(channelID string, candidate CallICECandidateInit) {
	m.PublishLocal(TopicCallLoopbackPrefix+channelID, "", CallLoopbackICEPayload{
		Type:      CallTypeLoopbackICE,
		ChannelID: channelID,
		Candidate: candidate,
	})
}

File: internal/ui/assets/js/mq/topics.js
/**
 * mq/topics.js — MQ topic registry.
 *
 * Defines all known topic strings as constants and exposes typed subscribe/send
 * helpers on Goop.mq. Mirrors internal/mq/topics.go — keep both in sync.
 *
 * Requires mq/base.js to be loaded first.
 *
 * ── Topic map ────────────────────────────────────────────────────────────────
 *
 *   peer:announce             Go → browser   peer metadata update (PublishLocal)
 *   peer:gone                 Go → browser   peer pruned (PublishLocal)
 *
 *   call:{channelID}          P2P + local    call signaling (see CALL_TYPES below)
 *   call:loopback:{channelID} Go → browser   LocalPC ICE candidates (Phase 4, PublishLocal)
 *
 *   group:{groupID}:{type}    P2P            join, welcome, members, msg, leave, close, ping, pong
 *   group.invite              P2P            group invitation (not scoped to groupID)
 *
 *   listen:{groupID}:state    Go → browser   listen state (PublishLocal)
 *   chat                      P2P            direct peer message
 *   chat.broadcast            P2P broadcast  message to all peers
 *   log:mq                    Go → browser   MQ event log entry (PublishLocal)
 *
 * ── Call signaling protocol ───────────────────────────────────────────────────
 *
 * All call signals travel on topic "call:{channelID}" with a "type" field:
 *
 *   caller                               callee
 *   ────────────────────────────────────────────────────────────────────────
 *   sendCallRequest ───────────────────► (incoming call modal shown)
 *                   ◄─────────────────── sendCallAck   (on accept)
 *   sendCallOffer   ───────────────────►
 *                   ◄─────────────────── sendCallAnswer
 *   sendCallICE ◄──────────────────────► sendCallICE  (trickle, both ways)
 *   sendCallHangup  ───────────────────► (or either side ends the call)
 *
 * Native-mode (Go/Pion): after call-request/ack, Go handles SDP offer/answer
 * and peer ICE internally. Browser receives media via WebSocket + WebM/MSE
 * (/api/call/media/{channel}). Loopback ICE uses "call:loopback:{channelID}".
 */
(function () {
  "use strict";

  var mq = window.Goop.mq;

  // ── Topic constants ───────────────────────────────────────────────────────────
  mq.TOPICS = Object.freeze({
    PEER_ANNOUNCE:         "peer:announce",
    PEER_GONE:             "peer:gone",
    CALL_PREFIX:           "call:",          // + channelID
    CALL_LOOPBACK_PREFIX:  "call:loopback:", // + channelID (Go → browser, Phase 4)
    GROUP_PREFIX:          "group:",         // + groupID + ":" + type
    GROUP_INVITE:          "group.invite",
    LISTEN_PREFIX:         "listen:",        // + groupID + ":state"
    CHAT:                  "chat",
    CHAT_BROADCAST:        "chat.broadcast",
    LOG_MQ:                "log:mq",
  });

  // ── Call signal type constants ────────────────────────────────────────────────
  // Value of the "type" field inside every call:* payload. Mirrors Go CallType* consts.
  mq.CALL_TYPES = Object.freeze({
    REQUEST:      "call-request",  // caller → callee: initiate a call
    ACK:          "call-ack",      // callee → caller: accepted, SDP exchange starts
    OFFER:        "call-offer",    // caller → callee: SDP offer
    ANSWER:       "call-answer",   // callee → caller: SDP answer
    ICE:          "ice-candidate", // either → other: trickle ICE candidate
    HANGUP:       "call-hangup",   // either side: end the call
    LOOPBACK_ICE: "loopback-ice",  // Go → browser: LocalPC ICE candidate (Phase 4)
  });

  // ── Typed subscribe helpers ────────────────────────────────────────────────────
  // Each returns an unsubscribe function (same as Goop.mq.subscribe).

  /** peer:announce — peer metadata update from Go PeerTable bridge */
  mq.onPeerAnnounce = function (fn) { return mq.subscribe(mq.TOPICS.PEER_ANNOUNCE, fn); };

  /** peer:gone — peer pruned from in-memory PeerTable */
  mq.onPeerGone = function (fn) { return mq.subscribe(mq.TOPICS.PEER_GONE, fn); };

  /**
   * onCall(fn) — all call:* signaling from any channel.
   * fn(from, topic, payload, ack) — payload.type is one of mq.CALL_TYPES.
   * channelId = topic.slice(5) strips the "call:" prefix.
   */
  mq.onCall = function (fn) { return mq.subscribe(mq.TOPICS.CALL_PREFIX + "*", fn); };

  /**
   * onLoopbackICE(channelId, fn) — Go's LocalPC ICE candidates for Phase 4 loopback.
   * fn(candidate) — RTCIceCandidateInit: { candidate, sdpMid, sdpMLineIndex }
   * Returns an unsubscribe function.
   */
  mq.onLoopbackICE = function (channelId, fn) {
    return mq.subscribe(mq.TOPICS.CALL_LOOPBACK_PREFIX + channelId, function (from, topic, payload, ack) {
      ack();
      if (payload && payload.candidate) {
        try { fn(payload.candidate); } catch (_) {}
      }
    });
  };

  /** group:{groupID}:{type} — all group protocol messages */
  mq.onGroup = function (fn) { return mq.subscribe(mq.TOPICS.GROUP_PREFIX + "*", fn); };

  /** group.invite — incoming group invitation */
  mq.onGroupInvite = function (fn) { return mq.subscribe(mq.TOPICS.GROUP_INVITE, fn); };

  /** listen:{groupID}:state — listen state updates from Go */
  mq.onListen = function (fn) { return mq.subscribe(mq.TOPICS.LISTEN_PREFIX + "*", fn); };

  /** chat — direct P2P message */
  mq.onChat = function (fn) { return mq.subscribe(mq.TOPICS.CHAT, fn); };

  /** chat.broadcast — broadcast message to all peers */
  mq.onChatBroadcast = function (fn) { return mq.subscribe(mq.TOPICS.CHAT_BROADCAST, fn); };

  /** log:mq — MQ event log entry from Go */
  mq.onLogMQ = function (fn) { return mq.subscribe(mq.TOPICS.LOG_MQ, fn); };

  // ── Typed send helpers — call protocol ───────────────────────────────────────

  /**
   * sendCall(peerId, channelId, payload) → Promise
   * Low-level: sends any call:* payload. Use the typed helpers below where possible.
   */
  mq.sendCall = function (peerId, channelId, payload) {
    return mq.send(peerId, mq.TOPICS.CALL_PREFIX + channelId, payload);
  };

  /**
   * sendCallRequest(peerId, channelId, constraints) → Promise
   * caller → callee: invite to a call.
   * constraints: optional browser-mode getUserMedia constraints { video, audio }.
   */
  mq.sendCallRequest = function (peerId, channelId, constraints) {
    var p = { type: mq.CALL_TYPES.REQUEST };
    if (constraints) p.constraints = constraints;
    return mq.sendCall(peerId, channelId, p);
  };

  /**
   * sendCallAck(peerId, channelId) → Promise
   * callee → caller: call accepted; triggers caller to create and send SDP offer.
   */
  mq.sendCallAck = function (peerId, channelId) {
    return mq.sendCall(peerId, channelId, { type: mq.CALL_TYPES.ACK });
  };

  /**
   * sendCallOffer(peerId, channelId, sdp) → Promise
   * caller → callee: WebRTC SDP offer (after receiving ack).
   */
  mq.sendCallOffer = function (peerId, channelId, sdp) {
    return mq.sendCall(peerId, channelId, { type: mq.CALL_TYPES.OFFER, sdp: sdp });
  };

  /**
   * sendCallAnswer(peerId, channelId, sdp) → Promise
   * callee → caller: WebRTC SDP answer.
   */
  mq.sendCallAnswer = function (peerId, channelId, sdp) {
    return mq.sendCall(peerId, channelId, { type: mq.CALL_TYPES.ANSWER, sdp: sdp });
  };

  /**
   * sendCallICE(peerId, channelId, candidate) → Promise
   * either → other: trickle ICE candidate.
   * candidate: RTCIceCandidateInit { candidate, sdpMid, sdpMLineIndex }
   */
  mq.sendCallICE = function (peerId, channelId, candidate) {
    return mq.sendCall(peerId, channelId, { type: mq.CALL_TYPES.ICE, candidate: candidate });
  };

  /**
   * sendCallHangup(peerId, channelId) → Promise
   * either side: end the call.
   */
  mq.sendCallHangup = function (peerId, channelId) {
    return mq.sendCall(peerId, channelId, { type: mq.CALL_TYPES.HANGUP });
  };

  // ── Typed send helpers — chat ────────────────────────────────────────────────

  /** sendChat(peerId, payload) → Promise — P2P direct message */
  mq.sendChat = function (peerId, payload) {
    return mq.send(peerId, mq.TOPICS.CHAT, payload);
  };

  /** broadcastChat(payload) → Promise — broadcast to all peers */
  mq.broadcastChat = function (payload) {
    return mq.broadcast(mq.TOPICS.CHAT_BROADCAST, payload);
  };

})();
